Numerous improvements

seanmil opened the pull request:
### Description

This PR contains a series of somewhat related improvements including:

* Aligning the communicator configuration options with the standard ones as provided in the packer-plugin-sdk
* Leverage the packer-plugin-sdk communicator setup routines to reduce custom code and enable additional features (e.g., SSH agent forwarding)
* Modify the port forwarding functionality to default to allocating a random ephemeral port
* Fix the port forwarding routine to gracefully handle connection errors, to permit communicator availability polling to work properly
* Add a way to disable port forwarding for direct communicator connections
* Add a wait_for_ip step which waits for the VM to report a guest IP and then provides that IP for possible use in later steps
* Add an optional wait_for_agent step which could be used to detect the phase of the installation where the Guest Agent becomes available

seanmil committed:
chore: Refactor communicator options

Refactor communicator-related options to leverage and align
with the provided functionality in the packer-plugin-sdk.

This change preserves backward compatibility with options
having different names but marks them as deprecated in favor
of the packer-plugin-sdk provided ones.

seanmil committed:
feat: Use packer-plugin-sdk communicator setup

Refactor the communicator setup to remove the custom setup
code and instead use the packer-plugin-sdk provided setup.

seanmil committed:
feat: Add wait_for_ip step

Modeled after the VMware plugin's IP wait, this adds a wait
on the availability of the VM's IP and then provides the IP
for use in other steps. If no forwarding is in use and
no IP/host was manually specified, the detected IP will
be used as the default.

seanmil committed:
feat: Overhaul port forwarding

Enable port forwarding to, by default, select and using a local
ephemeral port as the forwarding port. Also, adjust port
forwarding routines to permit connection failures gracefully
allowing for retries to support the communicator wait for
SSH/WinRM workflows.

seanmil committed:
feat: Add wait_for_agent step

If enabled, the builder will wait for the Guest Agent to be
reported as online prior to moving to the wait_for_ip step.

seanmil commented:
> Nice to see the packer-plugin-kubevirt evolving! Are the changes worth to adjust the included examples?

I took a quick look at the existing examples and, yes, I believe it would be worth adjusting them. I'll work on that and push an update. Thanks!

codingben reviewed (changes requested):
Looks good, thanks for taking care of this. I think examples should be updated as well and I want to make sure that the changes being tested against a real k8s cluster.

seanmil committed:
feat: Separate final datasource name from VM name

Allow the temporary VM name to be specified separately from
the desired resulting DataSource name. If not specified,
VMName will default to the final DataSource name.

seanmil committed:
feat: Enhance built media specification options

Add new configuration options media_content, media_label, and
keep_media to allow better flexibility and configurability
of the built ConfigMap-based media specification.

seanmil committed:
feat: Expose virtio_container config option

Enable users to specify their own container pull path for the
VirtIO Container image with the required Windows drivers.
Also, update the default container URL to the latest stable
version available.

seanmil committed:
docs: Mark installiation_wait_timeout as optional

With the addition of ip_wait_timeout the
installation_wait_timeout option should no longer be
marked as as required option.

seanmil committed:
docs: Minor documentation updates

seanmil committed:
feat: Add access_mode and volume_mode parameters

Expose the ability to set the accessMode and volumeMode used when
creating the PVC holding the temporary VM's rootdisk and the
PVC backing the resulting DataVolume built image.

seanmil committed:
test: Refactor and enhance resource tests

Refactor the create_bootablevolume and create_virtualmachine
tests to better support easy overrides of configuration values
to test different scenarios, and using this new refactor,
expand the testing coverage.

seanmil committed:
docs: Update and enhance example configurations

seanmil commented:
I just pushed a big update versus my first go at it. In order to properly build the whole set of systems I needed in my environment I needed to add more configuration options. I did include updated and expanded examples (adding SLES and Ubuntu), as well as significantly expanded unit test coverage vs. the last push.

seanmil committed:
feat: Add template_name configuration parameter

When specifying the "name" parameter in a simple HCL configuration,
using a static block definition, it gets correctly defined.

However, when using dynamic blocks the value set for "name" does
not get propagated into the builder configuration structure.

Exposing template_name as a separate parameter, and defaulting it
to "name" if defined, preserves backwards compatibility but also
allows the resulting template volume name to be properly set
when leveraging dynamic blocks in the configuration.

cskarby commented:
@codingben is there anything we can do to help make progress with the packer-plugin-kubevirt? E.g. split this PR into smaller pieces or anything else that makes it easier to review and make momentum?

codingben commented:
@cskarby I'm no longer active here as I'm contributing to KubeVirt projects. @tanmay-hc @anurag5sh Do you plan to review this PR?

EugenSusurrus commented:
Would be really nice to see these changes merged. Especially the ssh communicator ones, since the current kubevirt source, requires the ssh host to be set, This Makes the plugin virtually impossible to use in environments where the ip of the built VM is set dynamically.

Kind of sad that it takes so long to review and make it happen
